src/backend/distributed/commands/multi_copy.c

```diff
@@ -1013,42 +928,26 @@ OpenCopyTransactions(CopyStmt *copyStatement, ShardConnections *shardConnections
 			continue;
 		}
 
-		result = PQexec(connection, "BEGIN");
-		if (PQresultStatus(result) != PGRES_COMMAND_OK)
-		{
-			WarnRemoteError(connection, result);
-			failedPlacementList = lappend(failedPlacementList, placement);
-
-			PQclear(result);
-			continue;
-		}
-
-		PQclear(result);
+		ClaimConnectionExclusively(connection);
+		RemoteTransactionBeginIfNecessary(connection);
 		copyCommand = ConstructCopyStatement(copyStatement, shardConnections->shardId,
 											 useBinaryCopyFormat);
+		result = PQexec(connection->pgConn, copyCommand->data);
 
-		result = PQexec(connection, copyCommand->data);
 		if (PQresultStatus(result) != PGRES_COPY_IN)
 		{
-			WarnRemoteError(connection, result);
-			failedPlacementList = lappend(failedPlacementList, placement);
+			ReportConnectionError(connection, WARNING);
+			MarkRemoteTransactionFailed(connection, true);
 
 			PQclear(result);
+
+			/* failed placements will be invalidated by transaction machinery */
+			failedPlacementList = lappend(failedPlacementList, placement);
```

Comment:
Note that that presumably won't work for worker copies - presumably we should mark the connections critical there?